File: src/rename/scraper.py
import shutil
import requests
import xml.etree.ElementTree as ET
from pathlib import Path
from typing import Dict, Any, Optional, List, Set
from ..logger import logger
from ..config.config_manager import cm


class Scraper:
    def __init__(self):
        self.headers = {"User-Agent": "Mozilla/5.0"}
        self.tmdb_image_base_url = "https://image.tmdb.org/t/p/original"
        self.tmdb_image_dl_base = "https://image.tmdb.org/t/p/"
        self.tmdb_person_base = "https://www.themoviedb.org/person/"
        self.api_key = cm.get_config("api_key")

        self.url_cache: Dict[str, Path] = {}
        types = cm.get_config("scrape_image_types")
        self.allowed_types = set(types) if types else set()

    # -------------------------
    #  XML 辅助方法
    # -------------------------
    def _add_e(self, parent: ET.Element, tag: str, text: Any):
        if text is not None:
            s_text = str(text).strip()
            if s_text:
                elem = ET.SubElement(parent, tag)
                elem.text = s_text
                return elem
        return None

    def _add_uniqueid(self, parent: ET.Element, id_type: str, id_value: Any, is_default: bool = False):
        if id_value:
            elem = ET.SubElement(parent, "uniqueid")
            elem.text = str(id_value)
            elem.set("type", id_type)
            if is_default:
                elem.set("default", "true")
            else:
                elem.set("default", "false")

    def _save_nfo(self, root: ET.Element, save_path: Path):
        save_path.parent.mkdir(parents=True, exist_ok=True)
        # if save_path.exists(): return # 覆盖模式

        if hasattr(ET, "indent"):
            ET.indent(root, space="  ")

        try:
            tree = ET.ElementTree(root)
            tree.write(save_path, encoding="utf-8", xml_declaration=True)
            logger.info(f"[NFO生成] 已保存: {save_path.name}")
        except Exception as e:
            logger.warning(f"[NFO生成] 失败 {save_path.name}: {e}")

    # -------------------------
    #  核心元数据处理
    # -------------------------
    def _process_common_info(self, root: ET.Element, info: Dict[str, Any], is_movie: bool):
        tmdb_id = info.get("id")
        ext_ids = info.get("external_ids", {})
        imdb_id = ext_ids.get("imdb_id")
        tvdb_id = ext_ids.get("tvdb_id")

        self._add_e(root, "tmdbid", tmdb_id)
        self._add_uniqueid(root, "tmdb", tmdb_id, is_default=False)

        if tvdb_id:
            self._add_e(root, "tvdbid", tvdb_id)
            self._add_uniqueid(root, "tvdb", tvdb_id)

        if imdb_id:
            self._add_e(root, "imdbid", imdb_id)
            self._add_uniqueid(root, "imdb", imdb_id, is_default=True)

        overview = info.get("overview", "")
        self._add_e(root, "plot", overview)
        self._add_e(root, "outline", overview)

        self._process_credits(root, info)

        for genre in info.get("genres", []):
            self._add_e(root, "genre", genre.get("name"))

        self._add_e(root, "rating", info.get("vote_average"))
        self._add_e(root, "mpaa", self._get_certification(info, is_movie))

    def _process_credits(self, root: ET.Element, info: Dict[str, Any]):
        credits = info.get('credits', {})
        cast = credits.get('cast', [])
        for person in cast[:15]:
            actor_elem = ET.SubElement(root, 'actor')
            self._add_e(actor_elem, 'name', person.get('name'))
            self._add_e(actor_elem, 'type', 'Actor')
            self._add_e(actor_elem, 'role', person.get('character'))
            pid = person.get('id')
            self._add_e(actor_elem, 'tmdbid', pid)
            profile_path = person.get('profile_path')
            if profile_path:
                self._add_e(actor_elem, 'thumb', self.tmdb_image_base_url + profile_path)
            if pid:
                self._add_e(actor_elem, 'profile', f"{self.tmdb_person_base}{pid}")

        crew = credits.get('crew', [])
        for person in crew:
            job = person.get('job', '').lower()
            if job == 'director':
                self._add_e(root, 'director', person.get('name'))
            elif job in ['writer', 'screenplay', 'story', 'screenwriter']:
                self._add_e(root, 'credits', person.get('name'))

    def _get_certification(self, info: Dict[str, Any], is_movie: bool) -> str:
        try:
            if is_movie:
                releases = info.get('release_dates', {}).get('results', [])
                for country in releases:
                    if country.get('iso_3166_1') == "US":
                        for date_item in country.get('release_dates', []):
                            if date_item.get('certification'):
                                return date_item.get('certification')
            else:
                ratings = info.get('content_ratings', {}).get('results', [])
                for item in ratings:
                    if item.get('iso_3166_1') == "US":
                        return item.get('rating')
            return ""
        except:
            return ""

    # -------------------------
    #  图片下载
    # -------------------------
    def _smart_download(self, url_suffix: Optional[str], save_path: Path, image_type: str,
                        size: str = "original", allowed_types_override: Optional[Set[str]] = None):
        """
        :param allowed_types_override: 如果提供，则使用此集合检查；否则使用 self.allowed_types
        """
        # 确定当前的允许列表
        current_allowed = allowed_types_override if allowed_types_override is not None else self.allowed_types

        if not url_suffix or image_type not in current_allowed:
            return

        save_path.parent.mkdir(parents=True, exist_ok=True)
        if save_path.exists(): return

        full_url = f"{self.tmdb_image_dl_base}{size}{url_suffix}"

        if full_url in self.url_cache:
            if self.url_cache[full_url].exists():
                try:
                    shutil.copy2(self.url_cache[full_url], save_path)
                    return
                except:
                    pass

        try:
            resp = requests.get(full_url, headers=self.headers, timeout=15)
            if resp.status_code == 200:
                with open(save_path, "wb") as f:
                    f.write(resp.content)
                self.url_cache[full_url] = save_path
        except Exception:
            pass

    def _get_tmdb_images(self, tmdb_id: int, media_type: str) -> Dict[str, List[Dict]]:
        if not self.api_key: return {}
        url = f"https://api.themoviedb.org/3/{media_type}/{tmdb_id}/images"
        params = {"api_key": self.api_key, "include_image_language": "zh,cn,null,en"}
        try:
            resp = requests.get(url, params=params, headers=self.headers, timeout=10)
            if resp.status_code == 200: return resp.json()
        except:
            pass
        return {}

    # -------------------------
    #  辅助：提取 Config Override
    # -------------------------
    def _get_allowed_types(self, overrides: Optional[Dict]) -> Optional[Set[str]]:
        if overrides and 'scrape_image_types' in overrides:
            val = overrides['scrape_image_types']
            if val is not None:
                if isinstance(val, list):
                    return set(val)
                if isinstance(val, str):  # 如果配置里存的是逗号分隔字符串
                    return set([x.strip() for x in val.split(',') if x.strip()])
        return None

    # -------------------------
    #  入口方法
    # -------------------------
    def scrape_movie(self, save_path: Path, info: Dict[str, Any], config_overrides: Optional[Dict] = None):
        if not info: return
        tmdb_id = info.get("id")
        folder = save_path.parent

        allowed = self._get_allowed_types(config_overrides)

        images = self._get_tmdb_images(tmdb_id, "movie")
        p = images.get("posters", [])
        b = images.get("backdrops", [])

        self._smart_download(p[0]['file_path'] if p else info.get("poster_path"), folder / "poster.jpg", "poster",
                             allowed_types_override=allowed)
        self._smart_download(b[0]['file_path'] if b else info.get("backdrop_path"), folder / "fanart.jpg", "backdrop",
                             allowed_types_override=allowed)

        root = ET.Element('movie')
        self._add_e(root, "title", info.get("title"))
        self._add_e(root, "originaltitle", info.get("original_title"))
        self._add_e(root, "year", info.get("release_date", "")[:4])
        self._add_e(root, "premiered", info.get("release_date"))

        self._process_common_info(root, info, is_movie=True)
        self._save_nfo(root, save_path.with_suffix(".nfo"))

    def scrape_tv_show(self, work_path: Path, info: Dict[str, Any], config_overrides: Optional[Dict] = None):
        if not info: return
        tmdb_id = info.get("id")
        allowed = self._get_allowed_types(config_overrides)

        images = self._get_tmdb_images(tmdb_id, "tv")
        p = images.get("posters", [])
        b = images.get("backdrops", [])

        self._smart_download(p[0]['file_path'] if p else info.get("poster_path"), work_path / "poster.jpg", "poster",
                             allowed_types_override=allowed)
        self._smart_download(b[0]['file_path'] if b else info.get("backdrop_path"), work_path / "fanart.jpg",
                             "backdrop", allowed_types_override=allowed)

        root = ET.Element('tvshow')
        self._process_common_info(root, info, is_movie=False)

        self._add_e(root, "title", info.get("name"))
        self._add_e(root, "originaltitle", info.get("original_name"))
        self._add_e(root, "premiered", info.get("first_air_date"))
        self._add_e(root, "year", info.get("first_air_date", "")[:4])
        self._add_e(root, "season", "-1")
        self._add_e(root, "episode", "-1")

        self._save_nfo(root, work_path / "tvshow.nfo")

    def scrape_season(self, work_path: Path, season_number: int, info: Dict[str, Any], season_dir: Path = None,
                      config_overrides: Optional[Dict] = None):
        season_info = next((s for s in info.get("seasons", []) if s.get("season_number") == season_number), None)
        if not season_info: return

        allowed = self._get_allowed_types(config_overrides)
        poster = season_info.get("poster_path")

        self._smart_download(poster, work_path / f"season{season_number:02d}-poster.jpg", "poster",
                             allowed_types_override=allowed)
        if season_dir and season_dir != work_path:
            self._smart_download(poster, season_dir / "folder.jpg", "poster", allowed_types_override=allowed)

        target_dir = season_dir if season_dir else (work_path / f"Season{season_number}")
        if target_dir.exists():
            root = ET.Element('season')
            self._add_e(root, "title", season_info.get("name"))
            self._add_e(root, "season", season_number)
            self._add_e(root, "plot", season_info.get("overview"))
            self._add_e(root, "premiered", season_info.get("air_date"))
            self._add_e(root, "year", (season_info.get("air_date") or "")[:4])
            self._save_nfo(root, target_dir / "season.nfo")

    def scrape_episode(self, video_target: Path, info: Dict[str, Any], season_number: int, episode_num: int,
                       config_overrides: Optional[Dict] = None):
        season_info = next((s for s in info.get("seasons", []) if s.get("season_number") == season_number), None)
        if not season_info: return
        ep = next((e for e in season_info.get("episodes", []) if e.get("episode_number") == episode_num), None)
        if not ep: return

        allowed = self._get_allowed_types(config_overrides)

        if ep.get("still_path"):
            self._smart_download(ep.get("still_path"), video_target.with_name(f"{video_target.stem}-thumb.jpg"),
                                 "thumb", allowed_types_override=allowed)

        root = ET.Element('episodedetails')
        self._add_e(root, "title", ep.get("name"))
        self._add_e(root, "plot", ep.get("overview"))
        self._add_e(root, "season", season_number)
        self._add_e(root, "episode", episode_num)
        self._add_e(root, "aired", ep.get("air_date"))
        self._add_e(root, "rating", ep.get("vote_average"))

        eid = ep.get("id")
        self._add_e(root, "tmdbid", eid)
        self._add_uniqueid(root, "tmdb", eid, is_default=True)

        self._save_nfo(root, video_target.with_suffix(".nfo"))

File: go_app/internal/rename/scraper.go
package rename

import (
	"encoding/xml"
	"fmt"
	"io"
	"net/http"
	"os"
	"path/filepath"
	"strings"
	"time"
)

// ─────────────────────────────────────────────────────────────────────────────
// Scraper
// ─────────────────────────────────────────────────────────────────────────────

// Scraper downloads images and writes NFO sidecar files for Kodi/Jellyfin/Emby.
type Scraper struct {
	apiKey       string
	allowedTypes map[string]bool
	imageBase    string
	http         *http.Client
}

// NewScraper creates a Scraper with the given TMDB API key and allowed image
// type list (e.g. ["poster", "backdrop", "logo"]).
func NewScraper(apiKey string, allowedTypes []string) *Scraper {
	allowed := make(map[string]bool, len(allowedTypes))
	for _, t := range allowedTypes {
		allowed[strings.ToLower(t)] = true
	}
	return &Scraper{
		apiKey:       apiKey,
		allowedTypes: allowed,
		imageBase:    "https://image.tmdb.org/t/p/original",
		http:         &http.Client{Timeout: 30 * time.Second},
	}
}

// ─────────────────────────────────────────────────────────────────────────────
// Public entry points
// ─────────────────────────────────────────────────────────────────────────────

// ScrapeMovie writes movie.nfo and downloads poster/fanart for a movie file.
// savePath is the final target video file path; NFO and images are placed
// alongside it.
func (s *Scraper) ScrapeMovie(savePath string, info *TMDBMovieDetail) {
	if info == nil {
		return
	}
	folder := filepath.Dir(savePath)

	// Images
	s.downloadImage(info.PosterPath, filepath.Join(folder, "poster.jpg"), "poster", "w500")
	s.downloadImage(info.BackdropPath, filepath.Join(folder, "fanart.jpg"), "backdrop", "w1280")

	// NFO
	root := &movieNFO{
		Title:         info.Title,
		OriginalTitle: info.OriginalTitle,
		Year:          yearOf(info.ReleaseDate),
		Premiered:     info.ReleaseDate,
		Plot:          info.Overview,
		Outline:       info.Overview,
		Rating:        info.VoteAverage,
		MPAA:          movieCertification(info),
	}
	root.UniqueIDs = append(root.UniqueIDs, uniqueID{Type: "tmdb", Default: "true", Value: fmt.Sprintf("%d", info.ID)})
	if info.ExternalIDs.IMDbID != "" {
		root.UniqueIDs = append(root.UniqueIDs, uniqueID{Type: "imdb", Default: "false", Value: info.ExternalIDs.IMDbID})
	}
	for _, g := range info.Genres {
		root.Genres = append(root.Genres, g.Name)
	}
	root.Actors, root.Directors, root.Writers = extractCredits(info.Credits)

	nfoPath := strings.TrimSuffix(savePath, filepath.Ext(savePath)) + ".nfo"
	writeNFO(root, nfoPath)
}

// ScrapeTV writes tvshow.nfo and downloads show-level poster/fanart.
// workPath is the root show directory (e.g. /media/TV/ShowName (2020)).
func (s *Scraper) ScrapeTV(workPath string, info *TMDBTVDetail) {
	if info == nil {
		return
	}

	// Images
	s.downloadImage(info.PosterPath, filepath.Join(workPath, "poster.jpg"), "poster", "w500")
	s.downloadImage(info.BackdropPath, filepath.Join(workPath, "fanart.jpg"), "backdrop", "w1280")

	// Logo (first available)
	for _, logo := range info.Logos {
		if logo.ISO6391 == "en" || logo.ISO6391 == "" {
			s.downloadImage(logo.FilePath, filepath.Join(workPath, "logo.png"), "logo", "original")
			break
		}
	}

	root := &tvshowNFO{
		Title:         info.Name,
		OriginalTitle: info.OriginalName,
		Year:          yearOf(info.FirstAirDate),
		Premiered:     info.FirstAirDate,
		Plot:          info.Overview,
		Season:        "-1",
		Episode:       "-1",
		Rating:        0,
	}
	root.UniqueIDs = append(root.UniqueIDs, uniqueID{Type: "tmdb", Default: "false", Value: fmt.Sprintf("%d", info.ID)})
	if info.ExternalIDs.IMDbID != "" {
		root.UniqueIDs = append(root.UniqueIDs, uniqueID{Type: "imdb", Default: "true", Value: info.ExternalIDs.IMDbID})
	}
	if info.ExternalIDs.TVDbID > 0 {
		root.UniqueIDs = append(root.UniqueIDs, uniqueID{Type: "tvdb", Default: "false", Value: fmt.Sprintf("%d", info.ExternalIDs.TVDbID)})
	}
	for _, g := range info.Genres {
		root.Genres = append(root.Genres, g.Name)
	}
	root.Actors, root.Directors, root.Writers = extractCredits(info.Credits)

	writeNFO(root, filepath.Join(workPath, "tvshow.nfo"))
}

// ScrapeSeason writes season.nfo and downloads season poster.
// workPath is the show root; seasonNumber is the TMDB season number;
// seasonDir is the actual season directory on disk (may equal workPath).
func (s *Scraper) ScrapeSeason(workPath string, seasonNumber int, info *TMDBTVDetail, seasonDir string) {
	if info == nil {
		return
	}

	var seasonInfo *TMDBSeason
	for i := range info.Seasons {
		if info.Seasons[i].SeasonNumber == seasonNumber {
			seasonInfo = &info.Seasons[i]
			break
		}
	}
	if seasonInfo == nil {
		return
	}

	// Season poster alongside the show root
	posterName := fmt.Sprintf("season%02d-poster.jpg", seasonNumber)
	s.downloadImage(seasonInfo.PosterPath, filepath.Join(workPath, posterName), "poster", "w500")

	// Also write folder.jpg inside the season directory if it differs
	if seasonDir != "" && seasonDir != workPath {
		s.downloadImage(seasonInfo.PosterPath, filepath.Join(seasonDir, "folder.jpg"), "poster", "w500")
	}

	// season.nfo
	targetDir := seasonDir
	if targetDir == "" {
		targetDir = filepath.Join(workPath, fmt.Sprintf("Season%d", seasonNumber))
	}
	if _, err := os.Stat(targetDir); err == nil {
		root := &seasonNFO{
			Title:     seasonInfo.Name,
			Season:    seasonNumber,
			Plot:      seasonInfo.Overview,
			Premiered: seasonInfo.AirDate,
			Year:      yearOf(seasonInfo.AirDate),
		}
		writeNFO(root, filepath.Join(targetDir, "season.nfo"))
	}
}

// ScrapeEpisode writes an episode NFO and downloads the still image.
// videoTarget is the full path to the renamed video file.
func (s *Scraper) ScrapeEpisode(videoTarget string, info *TMDBTVDetail, seasonNumber, episodeNumber int) {
	if info == nil {
		return
	}

	var seasonInfo *TMDBSeason
	for i := range info.Seasons {
		if info.Seasons[i].SeasonNumber == seasonNumber {
			seasonInfo = &info.Seasons[i]
			break
		}
	}
	if seasonInfo == nil {
		return
	}

	var ep *TMDBEpisode
	for i := range seasonInfo.Episodes {
		if seasonInfo.Episodes[i].EpisodeNumber == episodeNumber {
			ep = &seasonInfo.Episodes[i]
			break
		}
	}
	if ep == nil {
		return
	}

	// Thumbnail
	if ep.StillPath != "" {
		thumbPath := strings.TrimSuffix(videoTarget, filepath.Ext(videoTarget)) + "-thumb.jpg"
		s.downloadImage(ep.StillPath, thumbPath, "thumb", "w300")
	}

	// Episode NFO
	root := &episodeNFO{
		Title:   ep.Name,
		Plot:    ep.Overview,
		Season:  seasonNumber,
		Episode: episodeNumber,
		Aired:   ep.AirDate,
		Rating:  ep.VoteAverage,
	}
	root.UniqueIDs = append(root.UniqueIDs, uniqueID{
		Type:    "tmdb",
		Default: "true",
		Value:   fmt.Sprintf("%d", ep.ID),
	})

	nfoPath := strings.TrimSuffix(videoTarget, filepath.Ext(videoTarget)) + ".nfo"
	writeNFO(root, nfoPath)
}

// ─────────────────────────────────────────────────────────────────────────────
// Image download helper
// ─────────────────────────────────────────────────────────────────────────────

func (s *Scraper) downloadImage(suffix, savePath, imageType, size string) {
	if suffix == "" || !s.allowedTypes[imageType] {
		return
	}
	if _, err := os.Stat(savePath); err == nil {
		return // already exists
	}
	if err := os.MkdirAll(filepath.Dir(savePath), 0o755); err != nil {
		return
	}

	url := fmt.Sprintf("https://image.tmdb.org/t/p/%s%s", size, suffix)
	resp, err := s.http.Get(url)
	if err != nil || resp.StatusCode != http.StatusOK {
		if resp != nil {
			resp.Body.Close()
		}
		return
	}
	defer resp.Body.Close()

	f, err := os.Create(savePath)
	if err != nil {
		return
	}
	defer f.Close()
	_, _ = io.Copy(f, resp.Body)
}

// ─────────────────────────────────────────────────────────────────────────────
// NFO XML structures
// ─────────────────────────────────────────────────────────────────────────────

type uniqueID struct {
	XMLName xml.Name `xml:"uniqueid"`
	Type    string   `xml:"type,attr"`
	Default string   `xml:"default,attr"`
	Value   string   `xml:",chardata"`
}

type actorXML struct {
	XMLName xml.Name `xml:"actor"`
	Name    string   `xml:"name"`
	Type    string   `xml:"type"`
	Role    string   `xml:"role,omitempty"`
	TMDbID  string   `xml:"tmdbid,omitempty"`
	Thumb   string   `xml:"thumb,omitempty"`
}

type movieNFO struct {
	XMLName       xml.Name   `xml:"movie"`
	Title         string     `xml:"title"`
	OriginalTitle string     `xml:"originaltitle"`
	Year          string     `xml:"year"`
	Premiered     string     `xml:"premiered"`
	Plot          string     `xml:"plot"`
	Outline       string     `xml:"outline"`
	Rating        float64    `xml:"rating"`
	MPAA          string     `xml:"mpaa,omitempty"`
	UniqueIDs     []uniqueID `xml:"uniqueid"`
	Genres        []string   `xml:"genre"`
	Actors        []actorXML
	Directors     []string `xml:"director"`
	Writers       []string `xml:"credits"`
}

type tvshowNFO struct {
	XMLName       xml.Name   `xml:"tvshow"`
	Title         string     `xml:"title"`
	OriginalTitle string     `xml:"originaltitle"`
	Year          string     `xml:"year"`
	Premiered     string     `xml:"premiered"`
	Plot          string     `xml:"plot"`
	Season        string     `xml:"season"`
	Episode       string     `xml:"episode"`
	Rating        float64    `xml:"rating"`
	MPAA          string     `xml:"mpaa,omitempty"`
	UniqueIDs     []uniqueID `xml:"uniqueid"`
	Genres        []string   `xml:"genre"`
	Actors        []actorXML
	Directors     []string `xml:"director"`
	Writers       []string `xml:"credits"`
}

type seasonNFO struct {
	XMLName   xml.Name `xml:"season"`
	Title     string   `xml:"title"`
	Season    int      `xml:"season"`
	Plot      string   `xml:"plot"`
	Premiered string   `xml:"premiered"`
	Year      string   `xml:"year"`
}

type episodeNFO struct {
	XMLName   xml.Name   `xml:"episodedetails"`
	Title     string     `xml:"title"`
	Plot      string     `xml:"plot"`
	Season    int        `xml:"season"`
	Episode   int        `xml:"episode"`
	Aired     string     `xml:"aired"`
	Rating    float64    `xml:"rating"`
	UniqueIDs []uniqueID `xml:"uniqueid"`
}

// ─────────────────────────────────────────────────────────────────────────────
// NFO write helper
// ─────────────────────────────────────────────────────────────────────────────

func writeNFO(v interface{}, path string) {
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		return
	}

	data, err := xml.MarshalIndent(v, "", "  ")
	if err != nil {
		return
	}

	content := []byte(xml.Header + string(data) + "\n")
	_ = os.WriteFile(path, content, 0o644)
}

// ─────────────────────────────────────────────────────────────────────────────
// Helpers
// ─────────────────────────────────────────────────────────────────────────────

func yearOf(dateStr string) string {
	if len(dateStr) >= 4 {
		return dateStr[:4]
	}
	return ""
}

func movieCertification(info *TMDBMovieDetail) string {
	for _, country := range info.ReleaseDates.Results {
		if country.ISO31661 == "US" {
			for _, rd := range country.ReleaseDates {
				if rd.Certification != "" {
					return rd.Certification
				}
			}
		}
	}
	return ""
}

func extractCredits(credits TMDBCredits) (actors []actorXML, directors []string, writers []string) {
	limit := 15
	if len(credits.Cast) < limit {
		limit = len(credits.Cast)
	}
	for _, person := range credits.Cast[:limit] {
		thumb := ""
		if person.ProfilePath != "" {
			thumb = "https://image.tmdb.org/t/p/original" + person.ProfilePath
		}
		actors = append(actors, actorXML{
			Name:   person.Name,
			Type:   "Actor",
			Role:   person.Character,
			TMDbID: fmt.Sprintf("%d", person.ID),
			Thumb:  thumb,
		})
	}
	for _, person := range credits.Crew {
		switch strings.ToLower(person.Job) {
		case "director":
			directors = append(directors, person.Name)
		case "writer", "screenplay", "story", "screenwriter":
			writers = append(writers, person.Name)
		}
	}
	return
}
